File: swift/713_SubarrayProductLessThanK.swift
class Solution {
    func numSubarrayProductLessThanK(_ nums: [Int], _ k: Int) -> Int {
        var count: Int = 0, trailIdx: Int = 0
        var currProduct: Int = 1

        for (i, n) in nums.enumerated() {
            currProduct *= n
            while currProduct >= k && trailIdx <= i {
                currProduct /= nums[trailIdx]
                trailIdx += 1
            }
            count += (i - trailIdx + 1)
        }
        return count
    }
}

File: java/713_SubarrayProductLessThanK.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int product = 1, trail = 0, count = 0;
        for (int i = 0; i < nums.length; ++i) {
            product *= nums[i];
            while (product >= k && trail <= i)
                product /= nums[trail++];
            count += (i - trail + 1);
        }
        return count;
    }
}
